File: strategy/signal/di_cross.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    RandomParameter,
)
from strategy.signal.base import Signal, SignalType


@dataclass(frozen=True)
class DiCrossSignal(Signal):
    type: SignalType = SignalType.DiCross
    period: Parameter = RandomParameter(10.0, 15.0, 1.0)
    signal_period: Parameter = RandomParameter(4.0, 8.0, 1.0)

File: ta_lib/strategies/signal/src/di_cross.rs
use base::prelude::*;
use core::prelude::*;
use momentum::di;

pub struct DICrossSignal {
    smooth_type: Smooth,
    period: usize,
    signal_period: usize,
}

impl DICrossSignal {
    pub fn new(smooth_type: Smooth, period: f32, signal_period: f32) -> Self {
        Self {
            smooth_type,
            period: period as usize,
            signal_period: signal_period as usize,
        }
    }
}

impl Signal for DICrossSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period, self.signal_period)
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let di = di(&data.close, self.smooth_type, self.period);
        let signal_line = di.smooth(self.smooth_type, self.signal_period);

        (di.cross_over(&signal_line), di.cross_under(&signal_line))
    }
}
